src/python/grpcio_tests/tests_aio/unit/client_stream_unary_interceptor_test.py

```diff
@@ -0,0 +1,531 @@
+# Copyright 2020 The gRPC Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import asyncio
+import logging
+import unittest
+import datetime
+
+import grpc
+
+from grpc.experimental import aio
+from tests_aio.unit._constants import UNREACHABLE_TARGET
+from tests_aio.unit._common import inject_callbacks
+from tests_aio.unit._test_server import start_test_server
+from tests_aio.unit._test_base import AioTestBase
+from tests.unit.framework.common import test_constants
+from src.proto.grpc.testing import messages_pb2, test_pb2_grpc
+
+_SHORT_TIMEOUT_S = 1.0
+
+_NUM_STREAM_REQUESTS = 5
+_REQUEST_PAYLOAD_SIZE = 7
+_RESPONSE_INTERVAL_US = int(_SHORT_TIMEOUT_S * 1000 * 1000)
+
+
+class _CountingRequestIterator:
+
+    def __init__(self, request_iterator):
+        self.request_cnt = 0
+        self._request_iterator = request_iterator
+
+    async def _forward_requests(self):
+        async for request in self._request_iterator:
+            self.request_cnt += 1
+            yield request
+
+    def __aiter__(self):
+        return self._forward_requests()
+
+
+class _StreamUnaryInterceptorEmpty(aio.StreamUnaryClientInterceptor):
+
+    async def intercept_stream_unary(self, continuation, client_call_details,
+                                     request_iterator):
+        return await continuation(client_call_details, request_iterator)
+
+    def assert_in_final_state(self, test: unittest.TestCase):
+        pass
+
+
+class _StreamUnaryInterceptorWithRequestIterator(
+        aio.StreamUnaryClientInterceptor):
+
+    async def intercept_stream_unary(self, continuation, client_call_details,
+                                     request_iterator):
+        self.request_iterator = _CountingRequestIterator(request_iterator)
+        call = await continuation(client_call_details, self.request_iterator)
```

Comment:
Is `write` allowed within the interceptors?

With mixing-api-style prohibited, it might be hard to do so since the interceptors are designed to pass a request iterator. Adding support for `write` here will further complex the interceptor logic. Maybe we should document that for client-side streaming RPCs `write` is not available in interceptors?

Reply:
`write` should not be used within the interceptors, but is perfectly usable from the caller even thought the `write` is being sent through an iterator behind the scenes.

There are some use cases within the iterator that they might have undesirable behaviors or at least unknown behaviors if they are not used in the right way.

That's the case of the `write, if the interceptor that has access o the full `call` object uses the `write` method will have as a side effect what you are saying unless the interceptor takes the responsibility to send everything that comes from the interceptor to `/dev/null`, so technically could be used.

The same will happen with the `__await__`, if the caller did not provide an iterator would never have the chance to write to the RPC since the interceptor would be blocking the call because of the `__await__`.

I can try to add some documentation, but TBH having the feeling that the support of the `write` (and the `read` as well`) is adding more friction and complexity in our implementation.

and most likely `__await__` neither.

Reply:
I agree about the complexity. Given current design, I don't have better ways to support `write` in interceptors.

There is one alternative I want to discuss, what if we don't link request iterator and `write` for users? If they want to intercept invocations of `write`, they have to update their interceptor implementation. For example, they could return a customized version of `Call` object that handles request iterator and `write` separately.

This means writing interceptors for streaming RPCs are slightly harder, but can reduce a large amount of logic needed in our implementation.

WDYT?

Reply:
> There is one alternative I want to discuss, what if we don't link request iterator and write for users? 

If we don't do that we will ask the interceptors to spread the logic in two different places, for when the user is using an iterator and when the user is using the write. With the current solution, the interceptor can provide a single point of logic.

In any case, your point is about the use case of the `write` within the interceptors which technically would be possible but would compromise the whole request iterator chain, since an interceptor using the chain will automatically circumvent that chain and will be writing directly to the call.

For that, we can either stress that out in the documentation, or we could basically return a narrowed version of the Call object with the `write` method removed.

WDYT?

Reply:
As you described, that solution requires duplication in users' interceptors. If the benefit is unclear, I think we should stick to this PR's design and add some documentation (or hide the `write` method). Documentation might be slightly better, since the `UsageError` already disabled using `write`.

Reply:
This is what I added [1]

[1] https://github.com/grpc/grpc/pull/22821/files#diff-173e8bd5d2219ba109f86237680068a4R177